Add -C option to overwrite config file parameters

ch-aurich opened the pull request:
add -C option to set/overwrite config file parameters (currently only for keyring:check-crl=false and keyring:check-purpose=codesign)

Fixes: #953 

<!--
Thank you for your pull request!

If this is your first pull request for RAUC, please read:
https://rauc.readthedocs.io/en/latest/contributing.html

Please describe what it changes, e.g. fixes a bug (and how), adds a feature, fixes documentation…

If you add a feature, please answer these questions:
- What do you use the feature for?
- How does RAUC benefit from the feature?
- How did you verify the feature works?
- If hardware is needed for the feature, which hardware is supported and which
  hardware did you test with?

Please also allow edits by maintainers, so we can apply minor fixes directly.
-->

<!--
In case your PR fixes an issue, please reference it in the next line, i.e.
Fixes: #[insert number without brackets here]
-->

jluebbe commented:
Thanks for your contribution! Why did you limit it to those two keyring options?

ch-aurich commented:
> Thanks for your contribution! Why did you limit it to those two keyring options?

I wanted to be able to verify the input, which is not an easy task to do for all the possible config parameters - hence the limitation to only the two parameters discussed in #953 and #729. But thinking about it again I would tend to leave that step (semantic check) out on this stage making it equivalent to the config file parsing where the g_key_file_load_from_data call takes everything that is syntactically correct. The semantic checks are done later nevertheless - so I could change it to accept all inputs as long as it is in the format SECTION:KEY=VALUE.

jluebbe commented on include/context.h:
```diff
@@ -22,6 +22,12 @@ typedef enum {
 	R_CONTEXT_CONFIG_MODE_REQUIRED, /* require config file */
 } RContextConfigMode;
 
+typedef struct {
+	gchar *overwrite_section;
+	gchar *overwrite_name;
+	gchar *overwrite_value;
+} ConfigFileOverwrite;
+
```
I'd use [override instead of overwrite](https://www.grammar.com/override_vs_overwrite). Also there's no need to repeat prefix all structure members with the same string.

```suggestion
typedef struct {
	gchar *section;
	gchar *name;
	gchar *value;
} ConfigFileOverride;
```

ee

jluebbe commented:
Yes, at this level we should just allow all inputs, as long as they fit the pattern.

Please also squash the commits to remove the intermediate changes and force-push this PR. A single commit is probably fine for this feature.

Could you also add a test case for the new code in `src/main.c`? Perhaps in [`test/test_info.py`](https://github.com/rauc/rauc/blob/master/test/test_info.py) by checking that `system:bundle-formats=verity` rejects plain bundles.

a3f commented:
Just some bikeshedding: Should we maybe call this `-o` (for option) instead? Being similar to SSH's `-o` may be easier to remember and I like the `option` term more than `--cmdconf`

ch-aurich commented:
> Just some bikeshedding: Should we maybe call this `-o` (for option) instead? Being similar to SSH's `-o` may be easier to remember and I like the `option` term more than `--cmdconf`

+1 from my side... reusing an existing pattern makes a ton of sense!

Just waiting for the confirmation from @jluebbe because you brought up the initial idea (https://github.com/rauc/rauc/issues/953#issuecomment-2346256501)

jluebbe commented:
I quite like the the correspondence between `-c` and `-C`. I don't think really need the long form, as that way the command line will grow really long. Otherwise perhaps `--confopt`?

The current version fails to compile.

Also please squash the commits as mentioned in https://github.com/rauc/rauc/pull/1542#issuecomment-2437685305.

ch-aurich commented:
> I quite like the the correspondence between `-c` and `-C`. I don't think really need the long form, as that way the command line will grow really long. Otherwise perhaps `--confopt`?

I changed it that way

> The current version fails to compile.

sorry, I now fixed it

> Also please squash the commits as mentioned in [#1542 (comment)](https://github.com/rauc/rauc/pull/1542#issuecomment-2437685305).

did that as well

ch-aurich commented:
something seems to be wrong with my new test in test_info.py - sorry, will fix that

ejoerns commented on src/main.c:
```diff
@@ -2646,6 +2646,41 @@ static void create_option_groups(void)
+	override->name = name;
+	override->value = value;
+	r_context_conf()->configoverride = g_list_append(r_context_conf()->configoverride, override);
+	return TRUE;
```
why do you return a boolean here?

ejoerns commented on src/main.c:
```diff
@@ -2646,6 +2646,41 @@ static void create_option_groups(void)
+	*equal_delimiter = 0;
+	equal_delimiter++;
+
+	return append_config_override(copy, colon_delimiter, equal_delimiter);
```
Not sure if this requires an extra method.

ch-aurich replied:
I put it in a separate function because initially I had the code to only accept known parameters and there I used it to not have duplicate code for different parameters. Now since we accept all parameters at this stage I can inline it again.

ejoerns commented on test/config_file.c:
```diff
@@ -1092,6 +1100,34 @@ check-purpose=codesign\n";
 	g_assert_true(config->keyring_check_crl);
 	g_assert_cmpstr(config->keyring_check_purpose, ==, "codesign-rauc");
+
+	g_free(pathname);
```
This does not really test the added functions but manually re-implements parts of their internal behavior. Maybe simply assigning allocated strings to each member of ConfigFileOverride is easier in any way?

jluebbe replied:
I'd be fine with that.

ch-aurich commented on src/main.c:
```diff
@@ -2646,6 +2646,41 @@ static void create_option_groups(void)
+	if (value == NULL) {
+		return FALSE;
+	}
+	gchar* copy = g_strdup(value);
```
Mh, any suggestions how else to do?

ejoerns replied:
It would maybe be okay when adding explicit comments in the code, the struct and the freeing code and if you manage to fix the memory leakage.

The alternative would be to use `g_strsplit()` as e.g. the slot parsing does. This gives an array of newly allocated strings, but two rounds are needed to split the items properly.

If you use autoptr cleanup, then you also don't need to care for the cleanup:

``` c
g_auto(GStrv) key_value_split = g_strsplit(value, "=", 2);

if (g_strv_length(key_value_split) != 2) {
	g_set_error(error, G_OPTION_ERROR, G_OPTION_ERROR_BAD_VALUE, "Missing '=' ... ");
	return FALSE;
}
```

You can collect the items in the end with autoptr stealing, like:

``` c
override->value = g_steal_pointer(&key_value_split[1]);
```

Autoptr cleanup could be used for your approach, too, maybe.

ch-aurich commented:
not sure why codecov does not notice the coverage at this point - from my point of view the test in pytest should cover the red lines as well:
![image](https://github.com/user-attachments/assets/0ee00f87-90cc-45ed-b557-bfb81b1d5202)

The missing parts that are really not tested are the error paths in main.c - should I add tests for that as well?

ejoerns reviewed (commented):
Parsing and memory handling looks really good to me now, Thank you :+1: 

Just a few final notes from my side, but nothing critical.

ejoerns commented:
> not sure why codecov does not notice the coverage at this point - from my point of view the test in pytest should cover the red lines as well: ![image](https://private-user-images.githubusercontent.com/428731/404336726-0ee00f87-90cc-45ed-b557-bfb81b1d5202.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzczNjM5OTgsIm5iZiI6MTczNzM2MzY5OCwicGF0aCI6Ii80Mjg3MzEvNDA0MzM2NzI2LTBlZTAwZjg3LTkwY2MtNDVlZC1iNTU3LWJmYjgxYjFkNTIwMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyMFQwOTAxMzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zOTlkOGNjMjE2NWEwODljMTcyYmMwMjI0OWI3NTdlYzIxZTllN2U1OWNlNWM0YzcyMWZjNjA4NmIwZGJhNzJhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.h1g0HOK84K_AjyJ_wFRxleaGKnQdTnYdXlDVQQdOTp4)

Mh, looks more like an off-by-one by codecov.

> The missing parts that are really not tested are the error paths in main.c - should I add tests for that as well?

As this should be quite straightforward, maybe add at least a test with a missing section name or so. This ensures that error propagation basically works.

ejoerns commented on src/main.c:
```diff
@@ -2654,6 +2682,7 @@ static void cmdline_handler(int argc, char **argv)
 	g_autoptr(GOptionContext) context = NULL;
 	GOptionEntry entries[] = {
 		{"conf", 'c', 0, G_OPTION_ARG_FILENAME, &confpath, "config file", "FILENAME"},
+		{"confopt", 'C', G_OPTION_FLAG_OPTIONAL_ARG, G_OPTION_ARG_CALLBACK, collect_config_values, "config settings that override parameters from the config file", "SECTION:KEY=VALUE" },
```
> It not only allows overriding what's in the config file but also allows adding new options that haven't been present before. But maybe that's a matter of perspective.
> 
> There should at least be a hint in the documentation, maybe a note box under docs/reference.rst?

Should have been more specific here. I meant a hint under the "System Configuration File" chapter that this option exists and how it can be overridden. But for me, your version is fine, too.

> Adding a parameter also requires adapting the manpage (`/rauc.1`).

Looks good, especially since it adds the details that `-h` output lacks.
